File: src/lib/agent-conversations-db.ts
import { randomUUID } from "crypto";
import { asc, eq } from "drizzle-orm";
import { getDb } from "./db";
import { agentConversations, agentConversationMessages } from "./schema";
import type { AgentConversationMeta, AgentConversationMessage } from "./agents/orchestrator/types";

export type { AgentConversationMeta, AgentConversationMessage };

/**
 * Orchestrator chat-thread persistence — Postgres `agent_conversations` /
 * `agent_conversation_messages`, mirroring canvas-db.ts's shape/conventions
 * (uuid ids, bigint-ms timestamps, app-supplied crypto.randomUUID()).
 */

type Row = typeof agentConversations.$inferSelect;

function rowToMeta(r: Row): AgentConversationMeta {
  return {
    id: r.id,
    projectId: r.projectId,
    name: r.name,
    kind: r.kind === "legacy" ? "legacy" : "chat",
    createdBy: r.createdBy ?? null,
    createdAt: r.createdAt,
    updatedAt: r.updatedAt,
  };
}

export async function listConversations(projectId: string): Promise<AgentConversationMeta[]> {
  const db = await getDb();
  const rows = await db
    .select()
    .from(agentConversations)
    .where(eq(agentConversations.projectId, projectId));
  return rows.map(rowToMeta);
}

export async function getConversation(id: string): Promise<AgentConversationMeta | undefined> {
  const db = await getDb();
  const rows = await db
    .select()
    .from(agentConversations)
    .where(eq(agentConversations.id, id))
    .limit(1);
  return rows[0] ? rowToMeta(rows[0]) : undefined;
}

async function insertConversation(
  projectId: string,
  name: string,
  kind: "chat" | "legacy",
  createdBy: string | null
): Promise<AgentConversationMeta> {
  const db = await getDb();
  const now = Date.now();
  const [row] = await db
    .insert(agentConversations)
    .values({ id: randomUUID(), projectId, name, kind, createdBy, createdAt: now, updatedAt: now })
    .returning();
  return rowToMeta(row);
}

export async function createConversation(
  projectId: string,
  name: string,
  createdBy: string | null
): Promise<AgentConversationMeta> {
  return insertConversation(projectId, name, "chat", createdBy);
}

/** Lazily creates the pinned "Old" thread the first time a project's list is
 *  fetched and none exists yet — same lazy-create pattern BoardSwitcher.tsx
 *  already uses for a project's first board. Idempotent per project. */
export async function ensureLegacyConversation(
  projectId: string
): Promise<AgentConversationMeta> {
  const existing = (await listConversations(projectId)).find((c) => c.kind === "legacy");
  if (existing) return existing;
  return insertConversation(projectId, "Old", "legacy", null);
}

export async function renameConversation(id: string, name: string): Promise<void> {
  const db = await getDb();
  await db
    .update(agentConversations)
    .set({ name, updatedAt: Date.now() })
    .where(eq(agentConversations.id, id));
}

/** Refuses to delete the legacy thread — callers must check `kind` first
 *  (the route does, before calling this) since it's the one thread the UI
 *  never offers a delete action for. */
export async function deleteConversation(id: string): Promise<void> {
  const db = await getDb();
  await db
    .delete(agentConversations)
    .where(eq(agentConversations.id, id));
  await db
    .delete(agentConversationMessages)
    .where(eq(agentConversationMessages.conversationId, id));
}

export async function listMessages(
  conversationId: string
): Promise<AgentConversationMessage[]> {
  const db = await getDb();
  const rows = await db
    .select()
    .from(agentConversationMessages)
    .where(eq(agentConversationMessages.conversationId, conversationId))
    .orderBy(asc(agentConversationMessages.createdAt));
  return rows.map((r) => ({
    id: r.id,
    conversationId: r.conversationId,
    role: r.role === "assistant" ? "assistant" : "user",
    content: r.content,
    toolTrace: r.toolTrace ?? null,
    createdAt: r.createdAt,
  }));
}

export async function appendMessage(
  conversationId: string,
  role: "user" | "assistant",
  content: string,
  toolTrace: { tool: string; args: unknown; result: unknown } | null = null
): Promise<AgentConversationMessage> {
  const db = await getDb();
  const now = Date.now();
  const [row] = await db
    .insert(agentConversationMessages)
    .values({ id: randomUUID(), conversationId, role, content, toolTrace, createdAt: now })
    .returning();
  await db
    .update(agentConversations)
    .set({ updatedAt: now })
    .where(eq(agentConversations.id, conversationId));
  return {
    id: row.id,
    conversationId: row.conversationId,
    role: row.role === "assistant" ? "assistant" : "user",
    content: row.content,
    toolTrace: row.toolTrace ?? null,
    createdAt: row.createdAt,
  };
}

File: backend/apps/agents/agent_conversations_db.py
"""Port of src/lib/agent-conversations-db.js — orchestrator chat-thread
persistence."""

import time
import uuid

from .models import AgentConversation, AgentConversationMessage


def _row_to_meta(c: AgentConversation) -> dict:
    return {
        "id": str(c.id),
        "projectId": str(c.project_id),
        "name": c.name,
        # Defensive fallback for rows written before this column existed.
        "agentKind": c.agent_kind if c.agent_kind == "video" else "image",
        "createdBy": str(c.created_by) if c.created_by else None,
        "createdAt": c.created_at,
        "updatedAt": c.updated_at,
    }


def _row_to_message(m: AgentConversationMessage) -> dict:
    return {
        "id": str(m.id),
        "conversationId": str(m.conversation_id),
        "role": m.role if m.role == "assistant" else "user",
        "content": m.content,
        "toolTrace": m.tool_trace,
        "createdAt": m.created_at,
    }


def list_conversations(project_id: str, agent_kind: str) -> list[dict]:
    return [_row_to_meta(c) for c in AgentConversation.objects.filter(project_id=project_id, agent_kind=agent_kind)]


def get_conversation(conversation_id: str) -> dict | None:
    c = AgentConversation.objects.filter(id=conversation_id).first()
    return _row_to_meta(c) if c else None


def create_conversation(project_id: str, agent_kind: str, name: str, created_by: str | None) -> dict:
    now = int(time.time() * 1000)
    c = AgentConversation.objects.create(
        id=uuid.uuid4(), project_id=project_id, agent_kind=agent_kind, name=name,
        created_by=created_by, created_at=now, updated_at=now,
    )
    return _row_to_meta(c)


def rename_conversation(conversation_id: str, name: str) -> None:
    AgentConversation.objects.filter(id=conversation_id).update(name=name, updated_at=int(time.time() * 1000))


def delete_conversation(conversation_id: str) -> None:
    AgentConversation.objects.filter(id=conversation_id).delete()
    AgentConversationMessage.objects.filter(conversation_id=conversation_id).delete()


def list_messages(conversation_id: str) -> list[dict]:
    return [
        _row_to_message(m)
        for m in AgentConversationMessage.objects.filter(conversation_id=conversation_id).order_by("created_at")
    ]


def attach_generated_item(message_id: str, item_id: str) -> dict | None:
    """Attaches the id of the GenerationItem a message's tool call
    actually produced. No-op (returns None) if the message has no
    toolTrace to attach onto."""
    existing = AgentConversationMessage.objects.filter(id=message_id).first()
    if not existing or not existing.tool_trace:
        return None
    tool_trace = {**existing.tool_trace, "generatedItemId": item_id}
    AgentConversationMessage.objects.filter(id=message_id).update(tool_trace=tool_trace)
    existing.refresh_from_db()
    return _row_to_message(existing)


def append_message(conversation_id: str, role: str, content: str, tool_trace: dict | None = None) -> dict:
    now = int(time.time() * 1000)
    m = AgentConversationMessage.objects.create(
        id=uuid.uuid4(), conversation_id=conversation_id, role=role, content=content,
        tool_trace=tool_trace, created_at=now,
    )
    AgentConversation.objects.filter(id=conversation_id).update(updated_at=now)
    return _row_to_message(m)
